lib/spack/spack/solver/asp.py

```diff
@@ -1289,7 +1302,22 @@ def target_defaults(self, specs):
         # TODO: investigate this.
         best_targets = set([uarch.family.name])
         for compiler in sorted(compilers):
-            supported = self._supported_targets(compiler, compatible_targets)
+            supported = self._supported_targets(
+                compiler.name, compiler.version, compatible_targets
+            )
+
+            # If we can't find supported targets it may be due to custom
+            # versions in the spec, e.g. gcc@foo. Try to match the
+            # real_version from the compiler object to get more accurate
+            # results.
+            if not supported:
+                compiler_obj = spack.compilers.compilers_for_spec(compiler)
+                compiler_obj = compiler_obj[0]
+                supported = self._supported_targets(
+                    compiler.name,
+                    compiler_obj.real_version,
+                    compatible_targets
+                )
```

Comment:
Is this necessary given that that the change to compiler_defaults() fixes the issue? This emits incorrect compiler_supports_target() facts if pointed at a cross-compiler. For instance if I set:

```yaml
- compiler:
    spec: gcc@foo
    operating_system: centos8
    paths:
      cc: /home/aweits/GIT/UPSTREAMSPACK/spack/opt/spack/linux-centos7-ivybridge/gcc-7.4.0/gcc-10.1.0-bhucsu2st2ljkksvusydawc7pcun446v/bin/powerpc64-linux-gnu-gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    modules: 'None'
    target: ppc
```

and do: 

```spack -m solve --show all a %gcc@foo target=ppc```

I will see:
```console
compiler_supports_target("gcc","foo","skylake_avx512").
compiler_supports_target("gcc","foo","x86_64").
compiler_supports_target("gcc","foo","skylake").
...etc...
```

Reply:
Ok I see now without this we won't pick up the microarchitectures - the fact that they're incorrect for a cross-compiler is a different problem entirely.

Reply:
@aweits It's an "optimization" see #20126